specs/ProgrammaticSaveAs.md

```diff
@@ -0,0 +1,267 @@
+                    MessageBox(
+                        m_appWindow->GetMainWindow(),
+                        (L"Save As " + saveAsResultString[result]).c_str(), L"Info", MB_OK);
+                }
+                return S_OK;
+            })
+            .Get());
+    return true;
+}
+//! [ProgrammaticSaveAs]
+```
+# API Details
+## Win32 C++
+```c++
+/// Specifies save as type selection options for `ICoreWebView2Staging20`,
+/// used in `SaveAsRequestedEventArgs`
+///
+/// When the source is a html page, allows to select `HTML_ONLY`,
+/// `SINGLE_FILE`, `COMPLETE`; when the source is a non-html,
+/// only allows to select `DEFAULT`; otherwise, will deny the download
+/// and return `COREWEBVIEW2_SAVE_AS_TYPE_NOT_SUPPORTED`.
+///
+/// The content type/format is a MIME type, indicated by the source
+/// server side and identified by the browser. It’s not related to the
+/// file’s type or extension. MIME type of `text/html`,
+/// `application/xhtml+xml` are considered as html page.
+[v1_enum] typedef enum COREWEBVIEW2_SAVE_AS_TYPE {
+  /// Default to save for a non-html content. If it is selected for a html
+  /// page, it’s same as HTML_ONLY option.
+  COREWEBVIEW2_SAVE_AS_TYPE_DEFAULT,
+  /// Save the page as html
+  COREWEBVIEW2_SAVE_AS_TYPE_HTML_ONLY,
+  /// Save the page as mhtml
+  COREWEBVIEW2_SAVE_AS_TYPE_SINGLE_FILE,
+  /// Save the page as html, plus, download the page related source files in
+  /// a folder
+  COREWEBVIEW2_SAVE_AS_TYPE_COMPLETE,
+} COREWEBVIEW2_SAVE_AS_TYPE;
+
+/// Status of a programmatic save as call, indicates the result
+/// for method `SaveContentAs`
+[v1_enum] typedef enum COREWEBVIEW2_SAVE_CONTENT_AS_RESULTS {
+  /// Programmatically open a system default save as dialog
+  COREWEBVIEW2_SAVE_AS_OPEN_SYSTEM_DIALOG,
+  /// Save as downloading not start as given an invalid path
+  COREWEBVIEW2_SAVE_AS_INVALID_PATH,
+  /// Save as downloading not start as given a duplicate filename and
+  /// replace file not allowed
+  COREWEBVIEW2_SAVE_AS_FILE_ALREADY_EXISTS,
+  /// Save as downloading not start as the `SAVE_AS_TYPE` selection not
+  /// supported because of the content MIME type or system limits
+  COREWEBVIEW2_SAVE_AS_TYPE_NOT_SUPPORTED,
+  /// Cancel the save as request
+  COREWEBVIEW2_SAVE_AS_CANCELLED,
+  /// Save as request complete, the downloading started
+  COREWEBVIEW2_SAVE_AS_STARTED
+} COREWEBVIEW2_SAVE_CONTENT_AS_RESULTS;
+
+
+[uuid(15e1c6a3-c72a-4df3-91d7-d097fbec3bfd), object, pointer_default(unique)]
+interface ICoreWebView2Staging20 : IUnknown {
+  /// Programmatically trigger a save as action for current content.
+  ///
+  /// Opens a system modal dialog by default. Returns COREWEBVIEW2_SAVE_AS_OPEN_SYSTEM_DIALOG.
+  /// If it was already opened, this method would not open another one
+  ///
+  /// If the silent save as option is enabled, won't open the system dialog, will
+  /// raise the `SaveAsRequested` event instead and process through its event args. The method can return
```

Comment:
What is the silent save as option? This paragraph sounds like the only way to get the SaveAsRequested event raised is to set the silent save as option but I don't see anything like that.

Reply:
I'm not sure if you mean having or not having an event handler is how you enable this option. We cannot vary our observable behavior based on if there are event handlers or not. That is, the end developer shouldn't notice a difference before or after subscribing to an event. Internally we can do things differently like not connect to the browser process or other sort of perf saving shortcuts, but we cannot change how the API behaves based on having an event handler or not.

Reply:
It looks like you indeed decide to show the file picker or not based on whether there are event handlers. We cannot do this. If the end developer adds an event handler that does nothing, they should see no difference from the WebView2 API as if they have no event handler. We can't change the default of showing a dialog based on if there are event handlers.